python-package/xgboost/training.py

```diff
@@ -109,22 +109,19 @@ def _train_internal(params, dtrain,
     else:
         raise ValueError(f'Unknown booster: {booster}')
 
-    num_groups = int(config['learner']['learner_model_param']['num_class'])
-    num_groups = 1 if num_groups == 0 else num_groups
     if bst.attr('best_score') is not None:
         bst.best_score = float(bst.attr('best_score'))
         bst.best_iteration = int(bst.attr('best_iteration'))
+        # num_class is handled internally
         bst.set_attr(
-            best_ntree_limit=str(
-                (bst.best_iteration + 1) * num_parallel_tree * num_groups
-            )
+            best_ntree_limit=str((bst.best_iteration + 1) * num_parallel_tree)
```

Comment:
It's not an exact revert since we also have fix for an old bug for `gblinear` with `ntree_limit`, which is valid.

Reply:
So the `best_ntree_limit` attribute is really a misnomer; it should have been called `best_iteration`. Is my understanding correct?

Reply:
The C++ layer considers `num_group` as it's a model parameter, and predictor `PredictBatch` has `GBTreeModel` as its argument.  But it doesn't consider `num_parallel_tree`, which is a training parameter instead of model parameter.  So the C++ function for `ntree_limit` is half implementation for `best_iteration`.

Reply:
So it's `best_iteration * num_parallel_tree` then? We should clearly document the meaning of `best_ntree_limit`, e.g.
```
Despite its name, the best_ntree_limit attribute is actually the product (best_iteration * num_parallel_tree).
If you set num_parallel_tree > 1, then best_ntree_limit won't be equal to the best boosting round.
```

Reply:
Just to clarify and reinforce:  The `best_ntree_limit` equals `num_parallel_tree * best_iteration`.  `num_class` is multiplied inside predictor.

When using classifier, the `best_ntree_limit` equals to `number of trees produced in best iteration / num_class`.

Reply:
The inplace prediction doesn't have this problem as it can use `best_iteration` directly.

Reply:
@hcho3 I will try to deprecate the attribute after setting it as a Python `@property` where we can throw proper warnings.

Reply:
Also, the sklearn model uses this attribute by default when running prediction.

Reply:
Added a short note on `train` function.